pkg/testutil/e2eutil/prometheus.go

```diff
@@ -277,6 +280,134 @@ func (p *Prometheus) Appender() tsdb.Appender {
 	return p.db.Appender()
 }
 
+// CreateEmptyBlock produces empty block like it was the case before fix: https://github.com/prometheus/tsdb/pull/374.
+// (Prometheus pre v2.7.0).
+func CreateEmptyBlock(dir string, blockDelay time.Duration, mint int64, maxt int64, extLset labels.Labels, resolution int64) (ulid.ULID, error) {
+	entropy := rand.New(rand.NewSource(time.Now().UnixNano()))
+	uid := ulid.MustNew(ulid.Now(), entropy)
+	if blockDelay != 0 {
+		var err error
+		uid, err = ulid.New(uint64(time.Now().Add(-blockDelay).Unix()*1000), entropy)
+		if err != nil {
+			return ulid.ULID{}, errors.Wrap(err, "create ulid")
+		}
+	}
+
+	if err := os.Mkdir(path.Join(dir, uid.String()), os.ModePerm); err != nil {
+		return ulid.ULID{}, errors.Wrap(err, "close index")
+	}
+
+	if err := os.Mkdir(path.Join(dir, uid.String(), "chunks"), os.ModePerm); err != nil {
+		return ulid.ULID{}, errors.Wrap(err, "close index")
+	}
+
+	w, err := index.NewWriter(context.Background(), path.Join(dir, uid.String(), "index"))
+	if err != nil {
+		return ulid.ULID{}, errors.Wrap(err, "new index")
+	}
+
+	if err := w.Close(); err != nil {
+		return ulid.ULID{}, errors.Wrap(err, "close index")
+	}
+
+	m := tsdb.BlockMeta{
+		Version: 1,
+		ULID:    uid,
+		MinTime: mint,
+		MaxTime: maxt,
+		Compaction: tsdb.BlockMetaCompaction{
+			Level:   1,
+			Sources: []ulid.ULID{uid},
+		},
+	}
+	b, err := json.Marshal(&m)
+	if err != nil {
+		return ulid.ULID{}, err
+	}
+
+	if err := ioutil.WriteFile(path.Join(dir, uid.String(), "meta.json"), b, os.ModePerm); err != nil {
+		return ulid.ULID{}, errors.Wrap(err, "saving meta.json")
+	}
+
+	if _, err = metadata.InjectThanos(log.NewNopLogger(), filepath.Join(dir, uid.String()), metadata.Thanos{
+		Labels:     extLset.Map(),
+		Downsample: metadata.ThanosDownsample{Resolution: resolution},
+		Source:     metadata.TestSource,
+	}, nil); err != nil {
+		return ulid.ULID{}, errors.Wrap(err, "finalize block")
+	}
+
+	return uid, nil
+}
+
+// CreateBlockWithBlockDelay writes a block with the given series and numSamples samples each.
+// Samples will be in the time range [mint, maxt)
+// Block ID will be created with a delay of time duration blockDelay.
+func CreateBlockWithBlockDelay(
+	ctx context.Context,
+	dir string,
+	series []labels.Labels,
+	numSamples int,
+	mint, maxt int64,
+	blockDelay time.Duration,
+	extLset labels.Labels,
+	resolution int64,
+) (id ulid.ULID, err error) {
+	blockID, err := createBlock(ctx, dir, series, numSamples, mint, maxt, extLset, resolution, false)
```

Comment:
It looks really similar to `CreateEmptyBlock`.

Can we reuse more code based on `numSamples`?

Reply:
I tried doing this, but getting an exception while calling tsdb Compactor.write function.
It throws exception when it calls `populateBlock` function.

Reply:
What kind of panic?

Reply:
so when we are creating a block without any samples, tsdb returns block id as ulid.ULID{} which is like 0000000000000000.
It is from this code: https://github.com/prometheus/prometheus/blob/master/tsdb/compact.go#L492

So we try to fetch block with id as 0, and thus fail to get its `meta.json` file and fail to populate Thanos labels and it panics.

Reply:
Then why not change the ulid?

Reply:
Anyway, I think it's fine for this PR. We can reuse more later on next one :+1:

Reply:
it doesn't return the ulid that it creates, also it doesn't create any block if numSamples = 0